File: app/src/main/java/ru/sukharev/myrecipes/addrecipe/AddRecipePresenter.java
package ru.sukharev.myrecipes.addrecipe;

import android.content.Context;

import ru.sukharev.myrecipes.model.IModel;
import ru.sukharev.myrecipes.model.RecipeRepository;
import ru.sukharev.myrecipes.pojo.Recipe;

/**
 * Presenter for add recipe view, using for adding new recipes with data which we got from add recipe view
 */
public class AddRecipePresenter implements AddRecipeContract.Presenter {

    private static AddRecipePresenter instance;

    private AddRecipeContract.View view;
    private IModel recipeModel;

    private int currentRating = 0;

    private AddRecipePresenter(Context context) {
        recipeModel = RecipeRepository.getInstance(context);
    }

    static void init(Context context, AddRecipeContract.View view){
        if (instance == null) {
            instance = new AddRecipePresenter(context.getApplicationContext());
        }
        instance.setView(view);
    }

    private void setView(AddRecipeContract.View view) {
        if (!view.equals(this.view)) {
            this.view = view;
            view.setPresenter(this);
        }
    }

    @Override
    public void start() {
        view.setRating(currentRating);
    }

    @Override
    public void addRecipe(String title, String desc) {
        final Recipe recipe = new Recipe.Builder()
                .setTitle(title)
                .setRating(currentRating)
                .setDescription(desc)
                .build();
        recipeModel.addRecipe(recipe, new IModel.AddRecipeCallback() {
            @Override
            public void onSuccess(long result) {
                view.onRecipeAdded();
            }

            @Override
            public void onFail() {
                view.onRecipeAddingError();
            }
        });
    }

    @Override
    public void setRating(int rating) {
        currentRating = rating;
        view.setRating(currentRating);
    }


}

File: app/src/main/java/ru/sukharev/myrecipes/addrecipe/AddRecipePresenter.kt
package ru.sukharev.myrecipes.addrecipe

import android.content.Context
import ru.sukharev.myrecipes.model.IModel
import ru.sukharev.myrecipes.model.RecipeRepository
import ru.sukharev.myrecipes.pojo.Recipe

/**
 * Presenter for add recipe view, using for adding new recipes with data which we got from add recipe view
 */
class AddRecipePresenter private constructor(context: Context) : AddRecipeContract.Presenter {

    private lateinit var view: AddRecipeContract.View
    private val recipeModel: IModel

    private var currentRating = 0

    init {
        recipeModel = RecipeRepository.getInstance(context)
    }

    private fun setView(view: AddRecipeContract.View) {
        if (view != this.view) {
            this.view = view
            view.setPresenter(this)
        }
    }

    override fun start() {
        view.setRating(currentRating)
    }

    override fun addRecipe(title: String, desc: String) {
        val recipe = Recipe.Builder()
                .setTitle(title)
                .setRating(currentRating)
                .setDescription(desc)
                .build()
        recipeModel.addRecipe(recipe, object : IModel.AddRecipeCallback {
            override fun onSuccess(result: Long) {
                view.onRecipeAdded()
            }

            override fun onFail() {
                view.onRecipeAddingError()
            }
        })
    }

    override fun setRating(rating: Int) {
        currentRating = rating
        view.setRating(currentRating)
    }

    companion object {

        private var instance: AddRecipePresenter? = null

        internal fun init(context: Context, view: AddRecipeContract.View){
            if (instance == null) {
                instance = AddRecipePresenter(context);
            }
            instance!!.setView(view)
        }
    }


}
